File: backend/GameApp.Domain/ValueObjects/Characters/ExplorerCharacter.cs
using GameApp.Domain.Entities.Items;
using GameApp.Domain.ValueObjects.Items;

namespace GameApp.Domain.ValueObjects.Characters;

// Concrete class for Explorer Character entity
public class ExplorerCharacter : Character
{
    private static readonly CharacterName CharacterName = new CharacterName("Explorer");
    private static readonly int DefaultMaxHealthPoints = 100;
    private static readonly int DefaultMaxFoodPoints = 140;
    private static readonly int DefaultMaxInventorySlots = 7;
    private static readonly int DefaultStartingMoney = 20;
    private static readonly int DefaultAttackSpeed = 2;
    private static readonly int DefaultAttackDamage = 3;


    // Character Ability
    private static readonly int NothingHappensScenesNeededToGetAbility = 3;
    private static readonly int ProbabilityOfRareItem = 15;
    private readonly int CurrentNothingHappensScenesVisited;
    


    // Default constructor
    public ExplorerCharacter()
        : base(CharacterName, DefaultMaxHealthPoints, DefaultMaxFoodPoints, DefaultMaxInventorySlots, DefaultStartingMoney, DefaultAttackSpeed, DefaultAttackDamage)
    {
        CurrentNothingHappensScenesVisited = 0;
    }

    // Restore constructor
    public ExplorerCharacter(int currentHealthPoints, int currentFoodPoints, int currentMoney, List<Item> inventoryList, int currentNothingHappensScenesVisited)
        : base(CharacterName, DefaultMaxHealthPoints, DefaultMaxFoodPoints, DefaultMaxInventorySlots, DefaultStartingMoney, DefaultAttackSpeed, DefaultAttackDamage,
                currentHealthPoints, currentFoodPoints, currentMoney, inventoryList)
    {
        CurrentNothingHappensScenesVisited = currentNothingHappensScenesVisited;
    }

    protected override Character CloneCharacter(int currentHealthPoints, int currentFoodPoints, int currentMoney, IReadOnlyList<Item> inventoryList)
    {
        List<Item> inventoryNewList = new List<Item>(inventoryList);
        return new ExplorerCharacter(currentHealthPoints, currentFoodPoints, currentMoney, inventoryNewList, CurrentNothingHappensScenesVisited);
    }

    // getter
    public int GetNothingHappensScenesToGetAbility() => NothingHappensScenesNeededToGetAbility;
    public int GetProbabilityOfRareItem() => ProbabilityOfRareItem;
    public int GetNothingHappensScenesVisited() => CurrentNothingHappensScenesVisited;

    // setter
    public ExplorerCharacter SetCurrentNothingHappensScenes(int currentNothingHappensScenes) => new ExplorerCharacter(GetCurrentHealthPoints(), GetCurrentFoodPoints(), GetCurrentMoney(), new List<Item>(GetInventoryList()), currentNothingHappensScenes);
    public ExplorerCharacter AddCurrentNothingHappensScenes()
    {
        int newValue = Math.Min(
            CurrentNothingHappensScenesVisited + 1,
            NothingHappensScenesNeededToGetAbility
        );

        return new ExplorerCharacter(
            GetCurrentHealthPoints(),
            GetCurrentFoodPoints(),
            GetCurrentMoney(),
            new List<Item>(GetInventoryList()),
            newValue
        );
    }

    // To string
    public override string ToString()
    {
        string inventoryStr = string.Join(", ", GetInventoryList().Select(i => i?.ToString() ?? "Empty"));
        return $"{GetName().GetName()} Explorer character(atq={GetAttackDamage()}, spd={GetAttackSpeed()}): " +
               $"HP={GetCurrentHealthPoints}/{GetMaxHealthPoints()}, Food={GetCurrentFoodPoints()}/{GetMaxFoodPoints()}, " +
               $"InventorySlots={GetMaxInventorySlots()}, Money={GetCurrentMoney()}, Inventory=[{inventoryStr}] currentNothingHappensScenes=[{CurrentNothingHappensScenesVisited}/{NothingHappensScenesNeededToGetAbility}]";
    }
}

File: frontend/gameapp-frontend/src/domain/value-objects/characters/explorer-character.ts
import { CharacterName } from "./character-name";
import { Character } from "./character";
import { Item } from "../../entities/items/item";

export class ExplorerCharacter extends Character {
    static CHARACTER_NAME = new CharacterName("Explorer");
    static DEFAULT_MAX_HEALTH = 100;
    static DEFAULT_MAX_FOOD = 140;
    static DEFAULT_MAX_SLOTS = 7;
    static DEFAULT_MONEY = 20;
    static DEFAULT_ATTACK_SPEED = 2;
    static DEFAULT_ATTACK_DAMAGE = 3;

    // Character Ability
    static NOTHING_HAPPENS_SCENES_NEEDED = 3;
    static PROBABILITY_OF_RARE_ITEM = 15;

    private readonly _currentNothingHappensScenes: number;

    constructor(
        currentHealthPoints: number | null = null,
        currentFoodPoints: number | null = null,
        currentMoney: number | null = null,
        inventoryList: Item[] | null = null,
        currentNothingHappensScenes: number = 0
    ) {
        super(
            ExplorerCharacter.CHARACTER_NAME,
            ExplorerCharacter.DEFAULT_MAX_HEALTH,
            ExplorerCharacter.DEFAULT_MAX_FOOD,
            ExplorerCharacter.DEFAULT_MAX_SLOTS,
            ExplorerCharacter.DEFAULT_MONEY,
            ExplorerCharacter.DEFAULT_ATTACK_SPEED,
            ExplorerCharacter.DEFAULT_ATTACK_DAMAGE,
            currentHealthPoints,
            currentFoodPoints,
            currentMoney,
            inventoryList
        );

        this._currentNothingHappensScenes = currentNothingHappensScenes;
    }

    cloneWith(
        currentHealthPoints: number,
        currentFoodPoints: number,
        currentMoney: number,
        inventoryList: Item[]
    ): ExplorerCharacter {
        return new ExplorerCharacter(
            currentHealthPoints,
            currentFoodPoints,
            currentMoney,
            [...inventoryList],
            this._currentNothingHappensScenes
        );
    }

    // getters
    getNothingHappensScenesToGetAbility(): number {
        return ExplorerCharacter.NOTHING_HAPPENS_SCENES_NEEDED;
    }

    getProbabilityOfRareItem(): number {
        return ExplorerCharacter.PROBABILITY_OF_RARE_ITEM;
    }

    getCurrentNothingHappensScenes(): number {
        return this._currentNothingHappensScenes;
    }

    canUseAbility(): boolean {
        return this._currentNothingHappensScenes >= ExplorerCharacter.NOTHING_HAPPENS_SCENES_NEEDED;
    }

    // setters
    setCurrentNothingHappensScenes(value: number): ExplorerCharacter {
        return new ExplorerCharacter(
            this._currentHealthPoints,
            this._currentFoodPoints,
            this._currentMoney,
            [...this._inventoryList],
            value
        );
    }

    addNothingHappensScene(): ExplorerCharacter {
        const newValue = Math.min(
            this._currentNothingHappensScenes + 1,
            ExplorerCharacter.NOTHING_HAPPENS_SCENES_NEEDED
        );
        return this.setCurrentNothingHappensScenes(newValue);
    }

    resetNothingHappensScenes(): ExplorerCharacter {
        return this.setCurrentNothingHappensScenes(0);
    }

    toString(): string {
        const inventoryStr = this._inventoryList
            .map(i => i?.toString() ?? "Empty")
            .join(", ");

        return `${this._name.toString()} Explorer character(atq=${this._attackDamage}, spd=${this._attackSpeed}): ` +
            `HP=${this._currentHealthPoints}/${this._maxHealthPoints}, ` +
            `Food=${this._currentFoodPoints}/${this._maxFoodPoints}, InventorySlots=${this._maxInventorySlots}, ` +
            `Money=${this._currentMoney}, Inventory=[${inventoryStr}] currentNothingHappensScenes=[${this._currentNothingHappensScenes}/${ExplorerCharacter.NOTHING_HAPPENS_SCENES_NEEDED}]`;
    }
}
